pypesto/sample/util.py

```diff
@@ -124,3 +127,252 @@ def bound_n_samples_from_env(n_samples: int):
+    max_iter: int
+        Maximum number of iterations. Default is 1000.
+    tol: float
+        Tolerance for convergence. Default is 1e-6.
+    """
+    from scipy import stats
+    from scipy.special import logsumexp
+
+    if result.sample_result is None:
+        raise ValueError("No samples available. Run sampling first.")
+    if not isinstance(result.problem.objective, AggregatedObjective):
+        raise ValueError("Objective must be an AggregatedObjective.")
+
+    # use Laplace approximation to get initial guess for p(y)
+    if initial_guess_log_evidence is None:
+        initial_guess_log_evidence = laplace_approximation_log_evidence(
+            problem=result.problem, x=result.optimize_result.x[0]
+        )
+    # extract posterior samples
+    burn_in = geweke_test(result)
+    posterior_samples = result.sample_result.trace_x[0, burn_in:]
+
+    # build proposal function from posterior samples
+    if n_posterior_samples_init is None:
+        n_posterior_samples_init = int(posterior_samples.shape[0] * 0.5)
+    # randomly select samples for calibration
+    calibration_index = np.random.choice(
+        np.arange(posterior_samples.shape[0]),
+        n_posterior_samples_init,
+        replace=False,
+    )
+    samples_calibration = posterior_samples[calibration_index]
+    # remove calibration samples from posterior samples
+    posterior_samples = posterior_samples[
+        [
+            j
+            for j in range(posterior_samples.shape[0])
+            if j not in calibration_index
+        ]
+    ]
+    # generate proposal samples and define proposal function
+    n_proposal_samples = posterior_samples.shape[0]
+    posterior_mean = np.mean(samples_calibration, axis=0)
+    posterior_cov = np.cov(samples_calibration.T)
+    if posterior_cov.size == 1:
+        # univariate case
+        proposal_samples = np.random.normal(
+            loc=posterior_mean,
+            scale=np.sqrt(posterior_cov),
+            size=n_proposal_samples,
+        )
+        proposal_samples = proposal_samples.reshape(-1, 1)
+    else:
+        # multivariate case
+        proposal_samples = np.random.multivariate_normal(
+            mean=posterior_mean, cov=posterior_cov, size=n_proposal_samples
+        )
+    log_proposal_fun = stats.multivariate_normal(
+        mean=posterior_mean, cov=posterior_cov
+    ).logpdf
+
+    # Compute the weights for the bridge sampling estimate
+    log_s1 = np.log(
+        posterior_samples.shape[0]
+        / (posterior_samples.shape[0] + n_proposal_samples)
+    )
+    log_s2 = np.log(
+        n_proposal_samples / (posterior_samples.shape[0] + n_proposal_samples)
+    )
+
+    # Start with the initial guess for p(y)
+    log_p_y = initial_guess_log_evidence
+
+    # Compute the log-likelihood, log-prior, and log-proposal for the posterior and proposal samples
+    # assumes that the objective function is the negative log-likelihood + negative log-prior
+    def log_likelihood_fun(x_array):
```

Comment:
this implicitely assumes a specific order in the objective functions. Would probably be better to test whether the objectives are appropriate ones?